core/subsurface-qt/DiveObjectHelper.cpp

```diff
@@ -208,176 +108,187 @@ QString DiveObjectHelper::gas() const
+		if (is_cylinder_used(d, i))
+			getCylinder << d->cylinder[i].type.description;
 	}
 	return getCylinder;
 }
 
-QStringList DiveObjectHelper::startPressure() const
+static QStringList getStartPressure(const dive *d)
 {
 	QStringList startPressure;
 	for (int i = 0; i < MAX_CYLINDERS; i++) {
-		if (is_cylinder_used(m_dive, i))
-			startPressure << getPressures(m_dive, i, START_PRESSURE);
+		if (is_cylinder_used(d, i))
+			startPressure << getPressures(d, i, START_PRESSURE);
 	}
 	return startPressure;
 }
 
-QStringList DiveObjectHelper::endPressure() const
+static QStringList getEndPressure(const dive *d)
 {
 	QStringList endPressure;
 	for (int i = 0; i < MAX_CYLINDERS; i++) {
-		if (is_cylinder_used(m_dive, i))
-			endPressure << getPressures(m_dive, i, END_PRESSURE);
+		if (is_cylinder_used(d, i))
+			endPressure << getPressures(d, i, END_PRESSURE);
 	}
 	return endPressure;
 }
 
-QStringList DiveObjectHelper::firstGas() const
+static QStringList getFirstGas(const dive *d)
 {
 	QStringList gas;
 	for (int i = 0; i < MAX_CYLINDERS; i++) {
-		if (is_cylinder_used(m_dive, i))
-			gas << get_gas_string(m_dive->cylinder[i].gasmix);
+		if (is_cylinder_used(d, i))
+			gas << get_gas_string(d->cylinder[i].gasmix);
 	}
 	return gas;
 }
 
+// Qt's metatype system insists on generating a default constructed object, even if that makes no sense.
+DiveObjectHelper::DiveObjectHelper()
+{
+}
+
+DiveObjectHelper::DiveObjectHelper(const struct dive *d) :
+	number(d->number),
+	id(d->id),
+	rating(d->rating),
+	visibility(d->visibility),
+	timestamp(d->when),
+	location(get_dive_location(d) ? QString::fromUtf8(get_dive_location(d)) : QString()),
+	gps(d->dive_site ? printGPSCoords(&d->dive_site->location) : QString()),
+	gps_decimal(format_gps_decimal(d)),
+	dive_site(QVariant::fromValue(d->dive_site)),
+	duration(get_dive_duration_string(d->duration.seconds, gettextFromC::tr("h"), gettextFromC::tr("min"))),
+	noDive(d->duration.seconds == 0 && d->dc.duration.seconds == 0),
+	depth(get_depth_string(d->dc.maxdepth.mm, true, true)),
+	divemaster(d->divemaster ? d->divemaster : QString()),
+	buddy(d->buddy ? d->buddy : QString()),
+	airTemp(get_temperature_string(d->airtemp, true)),
+	waterTemp(get_temperature_string(d->watertemp, true)),
+	notes(formatNotes(d)),
+	tags(get_taglist_string(d->tag_list)),
+	gas(formatGas(d)),
+	sac(formatSac(d)),
+	weightList(formatWeightList(d)),
+	weights(formatWeights(d)),
+	singleWeight(d->weightsystems.nr <= 1),
+	suit(d->suit ? d->suit : QString()),
+	cylinders(formatCylinders(d)),
+	cylinderObjects(makeCylinderObjects(d)),
+	tripId(d->divetrip ? QString::number((quint64)d->divetrip, 16) : QString()),
+	tripNrDives(d->divetrip ? d->divetrip->dives.nr : 0),
+	maxcns(d->maxcns),
+	otu(d->otu),
+	sumWeight(get_weight_string(weight_t { total_weight(d) }, true)),
+	getCylinder(formatGetCylinder(d)),
+	startPressure(getStartPressure(d)),
+	endPressure(getEndPressure(d)),
+	firstGas(getFirstGas(d))
+{
```

Comment:
So we use the constructor to render everything that we might need. Instead of rendering it on demand later. Now... how often do we call this constructor? Hmmm easy enough to test, I guess.

Reply:
> Now... how often do we call this constructor?

As you found out - often when showing the dive-list. This should be converted to "classical" Qt model/view semantics. Will do that.